### PR TITLE
Proxy Deployer fixes and optimisations

### PR DESCRIPTION
**Pull request by psorinionut**

### CODE REVIEW

File: contracts/proxy-deployer/src/config.rs [hunk 1]
```diff
@@ -60,6 +60,26 @@ pub trait ConfigModule {
...
+    }
+
+    #[only_owner]
+    #[endpoint(removeTemplateAddress)]
+    fn remove_template_address(&self, template_address: ManagedAddress) {
+        require!(
+            self.templates_list().contains(&template_address),
+            "Template address not found"
+        );
+        self.templates_list().swap_remove(&template_address);
```
Comment (dorin-iancu) [inline contracts/proxy-deployer/src/config.rs:80]:
```suggestion
        let _ = self.templates_list().swap_remove(&template_address);
```

File: contracts/proxy-deployer/src/contract_interactions.rs [hunk 2]
```diff
@@ -95,6 +107,66 @@ pub trait ContractInteractionsModule: config::ConfigModule + pause::PauseModule
...
+            "Contract address is not a SC"
+        );
+        require!(
+            !self.blacklisted_deployers_list().contains(&new_owner),
+            "New owner is blacklisted"
+        );
+        let mut caller = self.blockchain().get_caller();
+        if caller == self.blockchain().get_owner_address() {
+            require!(opt_orig_owner.is_some(), "Must provide original owner");
+            caller = opt_orig_owner.into_option().unwrap();
```
Comment (dorin-iancu) [inline contracts/proxy-deployer/src/contract_interactions.rs:131]: You already checked for some.

```suggestion
            caller = unsafe { opt_orig_owner.into_option().unwrap_unchecked() };
```

File: contracts/proxy-deployer/src/contract_interactions.rs [hunk 3]
```diff
@@ -95,6 +107,66 @@ pub trait ContractInteractionsModule: config::ConfigModule + pause::PauseModule
...
+        }
+        let contract_template_mapper = self.contract_template(&contract_address);
+        require!(!contract_template_mapper.is_empty(), "No template found");
+
+        let mut contract_processed = false;
+        let template_address = contract_template_mapper.take();
+        let deployer_template_addresses_mapper = self.deployer_template_addresses(&caller);
+        let mut deployer_template_addresses = deployer_template_addresses_mapper
+            .get(&template_address)
+            .unwrap();
```
Comment (dorin-iancu) [inline contracts/proxy-deployer/src/contract_interactions.rs:141]: Either use match and throw an error if it's empty, or if you're sure it's `Some()`, use `unwrap_unchecked`. `unwrap` function adds unnecessary checks and panics which bloat the wasm code.

File: contracts/proxy-deployer/src/contract_interactions.rs [hunk 4]
```diff
@@ -95,6 +107,66 @@ pub trait ContractInteractionsModule: config::ConfigModule + pause::PauseModule
...
+        }
+
+        require!(contract_processed, "Contract not found for deployer");
+        self.deployer_contracts(&caller)
+            .swap_remove(&contract_address);
+        if self.deployer_contracts(&caller).is_empty() {
+            self.deployers_list().swap_remove(&caller);
+        }
+
+        let () = self
```
Comment (dorin-iancu) [inline contracts/proxy-deployer/src/contract_interactions.rs:161]: I don't think you need to specify it has no return value?

Comment (psorinionut) [reply contracts/proxy-deployer/src/contract_interactions.rs:161]: It doesn't let me to compile without it, as it throws the error: `cannot satisfy `_: TopDecodeMulti``

File: contracts/proxy-deployer/src/contract_interactions.rs (hunk 3 above)
Comment (psorinionut) [reply contracts/proxy-deployer/src/contract_interactions.rs:141]: Done.

File: contracts/proxy-deployer/src/contract_interactions.rs (hunk 2 above)
Comment (psorinionut) [reply contracts/proxy-deployer/src/contract_interactions.rs:131]: Updated.

**Review by CostinCarabas — APPROVED** [review]

**Review by dorin-iancu — DISMISSED** [review]

File: contracts/proxy-deployer/src/contract_interactions.rs [hunk 5]
```diff
@@ -23,6 +23,15 @@ pub trait ContractInteractionsModule: config::ConfigModule + pause::PauseModule
             "Template address is not whitelisted"
         );
 
+        let ongoing_upgrade_operation_mapper = self.ongoing_upgrade_operation();
+        if !ongoing_upgrade_operation_mapper.is_empty() {
+            let ongoing_upgrade_operation = ongoing_upgrade_operation_mapper.get();
+            require!(
+                ongoing_upgrade_operation.template_address != template_address,
+                "There is an ongoing upgrade operation for this template address"
```
Comment (dorin-iancu) [inline contracts/proxy-deployer/src/contract_interactions.rs:31]: Error message seems confusing?

**Review by dorin-iancu — DISMISSED** [review]

**Review by dorin-iancu — APPROVED** [review]

**Review by CostinCarabas — APPROVED** [review]

**Pull request merged**